storage/innobase/dict/dict0upgrade.cc

```diff
@@ -937,18 +937,6 @@ bool dd_upgrade_table(THD *thd, const char *db_name, const char *table_name,
 
   dd_set_table_options(dd_table, ib_table);
 
-  /* Tables in encrypted general tablespace from PS-5.7 have
-  encryption attribute. Remove this attribute as tables in
-  general tablespace shouldn't have it. General tablespace
-  should have encryption attribute */
-  if (!dict_table_is_file_per_table(ib_table) &&
```

Comment:
ok, so it is reverse now.. :D  You should fix DD here.  ie. add encrypt_type to dd::Table here..

Reply:
As I understand, you are doing it in server

Reply:
The code from Oracle that sets the default encryption is here:
sql/dd/impl/bootstrap/bootstrapper.cc
-    // All the predefined tablespace are unencrypted (atleast for now).
-    tablespace->options().set("encryption", "N");
And the validation of encryption clause is done in sql/sql_table validate_table_encryption. From MySQL code I understood the intention was to implement all encryption stuff in server code, not in innodb

Reply:
Also scripts update encryption flags before the server start works on them - The encryption validation is done mysql_create_table_no_lock - before innodb is called - thus the encryption must be fixed in the server.

Reply:
may be true for regular case. During upgrade, server doesn't even know what InnoDB tablespaces are. It is SE (InnoDB) that reads SYS_TABLESPACES and register in DD. (dd_upgrade_tablespace())

Similarly, for table, .frm gives most details about table via TABLE and TABLE_SHARE and InnoDB DD has some more details about the table. It uses that too.. My point is that for upgrade, it is much easier and cleaner to register encryption attribute in DD.